Linear regression implementation.

param087 opened the pull request:
#4

param087 committed:
implemented linear regression

param087 committed:
updated linear regression notebook

param087 committed:
Delete Linear_Regression-checkpoint.ipynb

param087 committed:
updated linear regression notebook

param087 committed:
Merge branch 'linearRegression' of https://github.com/param087/swiftML into linearRegression

dan-zheng commented on Sources/swiftML/Helpers.swift:
```diff
@@ -0,0 +1,80 @@
+///   - b: Input tensor to find distance.
+///   - p: The distance metric to use for the tree, default set to 2.
+/// - Returns: Minkowski distance based on value of p.
+public func minkowskiDistance(_ a: Tensor<Float>, _ b: Tensor<Float>, _ p: Int ) -> Tensor<Float> {
```
How about including the argument label `p`, to distinguish it from the unlabeled arguments?
```suggestion
public func minkowskiDistance(_ a: Tensor<Float>, _ b: Tensor<Float>, p: Int) -> Tensor<Float> {
```

param087 replied:
That seems more appropriate in the case of reusability and understanding.

dan-zheng commented on Sources/swiftML/Helpers.swift:
```diff
@@ -0,0 +1,80 @@
+}
+
+  
+/// Return the minkowski distance based on value of p.
```
Documentation comment suggestions:
```suggestion
/// Returns the Minkowski distance between two tensors for the given distance metric `p`.
/// - Parameters:
///   - a: The first tensor.
///   - b: The second tensor.
///   - p: The order of the norm of the difference: `||a - b||_p`.
```

I don't feel a `/// - Returns` comment is necessary because the primary doc comment starts with `/// Returns`.

param087 replied:
I am thinking of specifying the data type/tensor structure in ` /// - Parameters` for type safety. What do you think about that?

param087 replied:
hey @dan-zheng,
For the documentation point of view, I was thinking of using [https://swift.org/documentation/api-design-guidelines/](https://swift.org/documentation/api-design-guidelines/)  and ([https://developer.apple.com/library/archive/documentat1ion/Xcode/Reference/xcode_markup_formatting_ref/index.html#//apple_ref/doc/uid/TP40016497-CH2-SW1](https://developer.apple.com/library/archive/documentation/Xcode/Reference/xcode_markup_formatting_ref/index.html#//apple_ref/doc/uid/TP40016497-CH2-SW1).) Is it ok? Or you have some specific guideline in mind?

rxwei replied:
You don't need to specify types in `- Parameter:` comments, since Swift is already statically and strongly typed.

param087 replied:
yeah, Before seeing documentation I was thinking of that but now I don't.

dan-zheng commented on Sources/swiftML/Helpers.swift:
```diff
@@ -0,0 +1,80 @@
+///   - b: Input tensor to find distance.
+///   - p: The distance metric to use for the tree, default set to 2.
+/// - Returns: Minkowski distance based on value of p.
+public func minkowskiDistance(_ a: Tensor<Float>, _ b: Tensor<Float>, _ p: Int ) -> Tensor<Float> {
```
What inputs are `minkowskiDistance` and `euclideanDistance` designed to take?
(1D tensors? tensors with the same shapes?)

Currently, without preconditions, the functions will work with tensors of different shapes, performing broadcasting. Consider an appropriate precondition, such as:
```swift
  precondition(a.shape == b.shape)
```

param087 replied:
Input to `minkowskiDistance` and `euclideanDistance` are 1D tensor of shape [n] . I will add appropriate precondition.

param087 committed:
added documentation suggestions

dan-zheng commented on Sources/swiftML/Helpers.swift:
```diff
@@ -0,0 +1,87 @@
+    uBasedDecision: Bool = true
+) -> (Tensor<Double>, Tensor<Double>) {
+  
+    var U = u
```
Please create mutable local variables named `u` and `v` instead of `U` and `V`. Swift supports this kind of aliasing.
```suggestion
    var u = u
    var v = v
```

param087 committed:
created mutable variables in Helpers.swift

dan-zheng committed:
Polish spelling and documentation comments.

- Use consistent capitalization and punctuation.
- Use consistent spelling for precondition messages.
- Add doc comments to stored properties.
- Use quasiquotes around code/math in doc comments.
- Change mutable local variable names to alias parameter names.
 Please enter the commit message for your changes. Lines starting

dan-zheng committed:
Polish Helpers.swift.

- Make helper functions be generic.
- Refactor `func matrixPseudoInverse` into a computed property
  `Tensor.pseudoinverse`.
- Use consistent spelling and capitalization.

dan-zheng commented:
I pushed some commits with changes.

Notable changes:
* Use consistent spelling in documentation comments and precondition messages.
* Change helper functions to be generic.
* Refactor `func matrixPseudoInverse` into a computed property `Tensor.pseudoinverse`, which is more idiomatic.

@param087: could you please review the changes to see if they make sense?

I saw that many PRs depend on `Helpers.swift` - could you please open a separate PR containing just `Helpers.swift`? Let's try to merge that PR ASAP to unblock your progress.

I have some more feedback regarding `LinearRegression.swift`, commenting soon.

dan-zheng committed:
Minor fixes.

dan-zheng committed:
Rename `iterations` to `iterationCount`.

dan-zheng commented on Sources/swiftML/LinearRegression.swift:
```diff
@@ -0,0 +1,124 @@
+            let ones = Tensor<Float>(ones: [data.shape[0], 1])
+            data = ones.concatenated(with: data, alongAxis: -1)
+        }
+        if self.gradientDescent == false {       
```
It seems that `class LinearRegression` currently has two different "strategies":
1. Gradient descent: which uses `fitIntercept`, `weights`, `learningRate`, `iterations`.
2. Singular value decomposition: which uses `fitIntercept`, `weights`.

Right now, the strategy is controlled by a flag `var gradientDescent: Bool`, but this is not particularly safe. `gradientDescent` may be `false`, but `learningRate` and `iterations` can still be accessed.

---

To make the design more safe, consider one of the following:
* Create a `protocol LinearRegression` and implement two separate conforming types: `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression`.
  * `protocol LinearRegression` defines requirement `func fit(data:labels:)` and provides default implementations for `func prediction(for:)` and `func score(data:labels:)` in an extension.
  * `class GradientDescentLinearRegression` and `class SingleValueDecompositionLinearRegression` conform to `protocol LinearRegression`, implement `func fit`, and define just the stored properties that they need. This way, `SingleValueDecompositionLinearRegression` won't be able to access `learningRate` or `iterationCount`.
* Keep a single `class LinearRegression`, but define a nested enum representing the "strategy".
  ```swift
  public class LinearRegresion {
      /// A linear regression strategy.
      enum Strategy {
          /// Gradient descent.
          case gradientDescent(learningRate: Float, iterationCount: Int)

          /// Singular value decomposition.
          case singularValueDecomposition
      }
      var strategy: Strategy // replaces `var gradientDescent: Bool`
      ...
  }
  ```
  * Then, create two factory methods: `static func gradientDescent(learningRate: Float, iterationCount: Int) -> LinearRegression` and `static func singleValueDecomposition() -> LinearRegression`. These will be used to construct instances of `LinearRegression`.

cc @rxwei

param087 replied:
@dan-zheng - I have started working on it.

param087 replied:
@dan-zheng - I have a quick question for you. What level of documentation should in protocol and implementation?

dan-zheng replied:
A 1-2 sentence summary per type declaration/function should be good!

For types/functions that implement well-known concepts (e.g. linear regression, matrix pseudoinverse, Minkowski distance), consider adding a reference link (e.g. to Wikipedia).

You can look at [tensorflow/swift-apis](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Loss.swift) as a reference for documentation: see [optimizer](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Optimizers/MomentumBased.swift) documentation.

param087 committed:
Restructured LinearRegression class to be more safe
- Created a protocol LinearRegression
- Implemented two separate conforming types: class GradientDescentLinearRegression and class SingleValueDecompositionLinearRegression.
- Updated LinearRegression notebook

param087 committed:
Remove svdFlip and replace with deterministicSvd

param087 committed:
Minor indentation fix

param087 committed:
Merge branch 'master' of https://github.com/param087/swiftML into linearRegression

param087 committed:
Added end-to-end linear regression examples
added support for multiple target values in linear Regression

param087 committed:
Added end-to-end linear regression examples
Corrected spelling of `SingleValueDecompositionLinearRegression` -> `SingularValueDecompositionLinearRegression`
Added support for multiple target values in linear regression

param087 committed:
minor indentation and notebook update.

param087 committed:
Merge branch 'master' of https://github.com/param087/swiftML into linearRegression
